src/compiler/moduleSpecifiers.ts

```diff
@@ -828,51 +836,82 @@ function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: s
-                    }
+        // conditional mapping
+        for (const key of getOwnKeys(exports as MapLike<unknown>)) {
+            if (key === "default" || conditions.indexOf(key) >= 0 || isApplicableVersionedTypesKey(conditions, key)) {
+                const subTarget = (exports as MapLike<unknown>)[key];
+                const result = tryGetModuleNameFromExportsOrImports(options, targetFilePath, packageDirectory, packageName, subTarget, conditions, mode);
+                if (result) {
+                    return result;
                 }
             }
         }
     }
     return undefined;
 }
 
+function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[]): { moduleFileToTry: string } | undefined {
+    if (typeof exports === "object" && exports !== null && !Array.isArray(exports) && allKeysStartWithDot(exports as MapLike<unknown>)) { // eslint-disable-line no-null/no-null
+        // sub-mappings
+        // 3 cases:
+        // * directory mappings (legacyish, key ends with / (technically allows index/extension resolution under cjs mode))
+        // * pattern mappings (contains a *)
+        // * exact mappings (no *, does not end with /)
+        return forEach(getOwnKeys(exports as MapLike<unknown>), k => {
+            const subPackageName = getNormalizedAbsolutePath(combinePaths(packageName, k), /*currentDirectory*/ undefined);
+            const mode = endsWith(k, "/") ? MatchingMode.Directory
+                : stringContains(k, "*") ? MatchingMode.Pattern
+                : MatchingMode.Exact;
+            return tryGetModuleNameFromExportsOrImports(options, targetFilePath, packageDirectory, subPackageName, (exports as MapLike<unknown>)[k], conditions, mode);
+        });
+    }
+    return tryGetModuleNameFromExportsOrImports(options, targetFilePath, packageDirectory, packageName, exports, conditions, MatchingMode.Exact);
+}
+
+function tryGetModuleNameFromPackageJsonImports(moduleFileName: string, sourceDirectory: Path, options: CompilerOptions, host: ModuleSpecifierResolutionHost, importMode: ResolutionMode) {
+    if (!host.readFile || !getResolvePackageJsonImports(options)) {
+        return undefined;
+    }
+
+    const ancestorDirectoryWithPackageJson = getNearestAncestorDirectoryWithPackageJson(host, sourceDirectory);
+    if (!ancestorDirectoryWithPackageJson) {
+        return undefined;
+    }
+    const packageJsonPath = combinePaths(ancestorDirectoryWithPackageJson, "package.json");
+    const cachedPackageJson = host.getPackageJsonInfoCache?.()?.getPackageJsonInfo(packageJsonPath);
+    if (typeof cachedPackageJson !== "object" && cachedPackageJson !== undefined) {
+        return undefined;
+    }
+    const packageJsonContent = cachedPackageJson?.contents.packageJsonContent || readJson(packageJsonPath, host as { readFile(fileName: string): string | undefined });

```

Comment:
Use `JSON.parse` for package.json files. `readJson` uses the TypeScript parser to construct an AST and then converts that to an object 😵 I’ve been meaning to audit existing uses of `readJson` and then rename it; I know some can be replaced.

Reply:
Yeah, I saw that 😅. I was aiming to consistent with the implementation of auto-import for the `exports` field. I was thinking maybe that should be left for a seperate PR to avoid more ad-hoc handling of package.json reading?

> Use JSON.parse for package.json files. readJson uses the TypeScript parser to construct an AST and then converts that to an object 😵 I’ve been meaning to audit existing uses of readJson and then rename it; I know some can be replaced.

I've actually checked btw and all the usages of `readJson` are for reading `package.json`s and the `readJsonOrUndefined` function that's used by `readJson` is only used for build info so all of them can really be replaced with `JSON.parse`

Reply:
I'm not sure that replacing everything with `JSON.parse` is generally correct, because it'll throw if the file has malformed JSON. The services project has a helper, `tryParseJson`, which catches this, and that's used by its `createPackageJsonInfo` (which then makes its package.json cache).

Reply:
Yes, `JSON.parse` should be in a try/catch, particularly when reading a local package.json as opposed to one in node_modules. It’s virtually impossible for a node_modules package.json to be malformed so I don’t care as much there. My point was only that we shouldn’t be creating TypeScript AST nodes and issuing parse diagnostics just to get an object representation of a JSON file.